templatised foundry deploy script

technophile-04 opened the pull request:
### Description : 

This will allow people to create extension's which adds new contract and also deployments for them. 

### To test : 
Switch to this branch then: 

1. Start cli : 
```
yarn dev
```
2. Scaffold new app, with eip-5792 extension: 

```
yarn cli -e technophile-04/eip-5792:foundry
```

3. Choose foundry as extension too

Here is extension [repo](https://github.com/technophile-04/eip-5792/tree/foundry/extension/packages)

--- 

Adding contracts and deployments in hardhat does not require any templatisation since we can create separate `01_deploy_..` file and main deploy script `99_deploy...` will consider all the deployments. 

But in foundry this is not the case [checkout this](https://github.com/scaffold-eth/scaffold-eth-2/pull/781#issuecomment-2018642266) for more details

rin-st commented:
Thanks Shiv, good job!

Trying to run
```
yarn cli -e technophile-04/eip-5792:foundry
```
and getting this error when deploying contracts (sorry for screenshots, adding as images because of colors)

<img width="779" alt="image" src="https://github.com/scaffold-eth/create-eth/assets/25638585/fac3156a-b549-4eca-bb83-96924820a5d3">

if I move `vm.stopBroadcast()` upper in `Deploy.s.sol`

```solidity
    vm.startBroadcast(deployerPrivateKey);

    YourContract yourContract = new YourContract(vm.addr(deployerPrivateKey));
    console.logString(string.concat("YourContract deployed at: ", vm.toString(address(yourContract))));

    vm.stopBroadcast();

    DeployExampleContract deployExampleContract = new DeployExampleContract();
    deployExampleContract.run();
```
I'm getting this error
<img width="1076" alt="image" src="https://github.com/scaffold-eth/create-eth/assets/25638585/f69908cd-4ab6-4946-8577-69c95eb7c599">

Looks like `DeployHelpers.s.sol` should create folder/files if they don't exist.

Upd. Added empty `deployments/31337.json` and deployment worked 🎉 . But looks like it should be changed in another pr.

rin-st commented on templates/extensions/foundry/packages/foundry/script/Deploy.s.sol.template.mjs:
```diff
@@ -11,17 +14,23 @@ contract DeployScript is ScaffoldETHDeploy {
             )
         );
+
+        ${deploymentsLogic.filter(Boolean).join("\n")}
```
should it be moved below `vm.stopBroadcast()`? See first error https://github.com/scaffold-eth/create-eth/pull/43#issuecomment-2145651951

technophile-04 replied:
Ohh yes fixed at https://github.com/scaffold-eth/create-eth/pull/43/commits/0c109be9ad2895222a09602487b08f540c627d5c thanks !

technophile-04 commented:
> Upd. Added empty deployments/31337.json and deployment worked 🎉 . But looks like it should be changed in another pr.

Ohh yeah thanks !, actually this bug was introduced in [#38](https://github.com/scaffold-eth/create-eth/pull/38/files#diff-c0c6f6da298e61626972856bdcade07e26609224b1d7578ec11ef638d083e4d1) and currently present `main`. 

In #38 we removed `.gitignore` from `foundry/deployments` and since `.gitignore` was only file present in `deployments` git removed the whole `deployments` dir and stopped tracking it. 

Fixed this at https://github.com/scaffold-eth/create-eth/pull/43/commits/8e2f18d224030dd65436aef0f5f44936b23429b5 by adding `deployments/31337.json` this makes sure that `deployments` dir is present when we run `yarn deploy`. We can't have empty `deployments` in cli template since git is not able to track it. But I think its fine to have `deployments/31337.json` already present since `31337` is any which way git ignored in scaffolded app and also it just a small file which mentions name of the network so its always deterministic

rin-st reviewed (approved):
Lgtm! Thank you!

technophile-04 commented:
Merging this, thanks !! 🙌
